src/sage/graphs/orientations.py

```diff
@@ -41,6 +41,233 @@
 from sage.graphs.digraph import DiGraph
 
 
+def acyclic_orientations(G):
+    r"""
+    Return an iterator over all acyclic orientations of an undirected graph `G`.
+
+    ALGORITHM:
+
+    The algorithm is based on a paper by Mathew B. Squire, presenting an efficient
+    algorithm for listing the acyclic orientations of a graph. The algorithm is
+    shown to require O(n) time per acyclic orientation generated, making it the
+    most efficient known algorithm for generating acyclic orientations.
+
+    The function uses a recursive approach to generate acyclic orientations of the
+    graph. It reorders the vertices and edges of the graph, creating a new graph
+    with updated labels. Then, it iteratively generates acyclic orientations by
+    considering subsets of edges and checking whether they form upsets in a
+    corresponding poset.
+
+    INPUT:
+
+    - ``G`` -- an undirected graph.
+
+    OUTPUT:
+
+    - An iterator over all acyclic orientations of the input graph.
+
+    .. NOTE::
+
+        The function assumes that the input graph is undirected and the edges are unlabelled.
+
+    EXAMPLES:
+
+    To count number acyclic orientations for a graph::
+
+        sage: G = Graph([(0, 3), (0, 4), (3, 4), (1, 3), (1, 2), (2, 3), (2, 4)])
+        sage: it = g.acyclic_orientations(G)
+        sage: len(list(it))
+        54
+
+    For C100 graph, we get first 10 orientations pretty quickly::
+
+        sage: from itertools import islice
+        sage: import timeit
+        sage: G_C100, start_time = graphs.CycleGraph(100), timeit.default_timer()
+        sage: it_C100, first_10_orientations = acyclic_orientations(G_C100), list(islice(acyclic_orientations(G_C100), 10))
+        sage: print(timeit.default_timer() - start_time < 0.05)
+        True
+
+    TESTS:
+
+    Acyclic orientations of a complete graph::
+
+        sage: G = graphs.CompleteGraph(5)
+        sage: it = g.acyclic_orientations(G)
+        sage: len(list(it))
+        120
+    """
+    from sage.rings.infinity import Infinity
+    from sage.combinat.subset import Subsets
+
+    def reorder_vertices(G):
```

Comment:
if the vertices of `G` are not labeled in range `0..n-1`, `vertex_labels` should be a mapping vertex to int.

Reply:
We are assuming that the edges are unlabelled 
<img width="735" alt="image" src="https://github.com/sagemath/sage/assets/111726832/c692fe0f-bcd8-48ca-956b-50f7a2ca5448">

Reply:
Yes, edges are unlabelled, but the vertex of a graph can be an integer, a string or any hashable object (frozenset, immutable graph, etc.). The current implementation is not working if a vertex is named `'abc'`.

Reply:
Got it! I have made the changes and added test cases for the same.